src-tauri/src/aux_data_structures/timeline.rs

```diff
@@ -0,0 +1,234 @@
+        let mut curr_timeline_id = 0;
+
+        let timeline_data_dir = config_fields.get("timeline_data_dir").unwrap();
+        let timeline_label_dir = config_fields.get("timeline_label_dir").unwrap();
+
+        let paths = fs::read_dir(timeline_data_dir).unwrap();
+        for path in paths {
+            let path = path.unwrap().path();
+            let filename = path.file_name().unwrap().to_str().unwrap();
+            let filename = filename.split('.').collect::<Vec<&str>>()[0];
+            let filename = filename.parse::<i32>().unwrap();
+            if filename > curr_timeline_id {
+                curr_timeline_id = filename;
+            };
+        }
+
+        curr_timeline_id += 1;
+
+        Timeline {
+            curr_snapshot: None,
+            label: 1,
+            curr_timeline_id,
+            curr_snapshot_id: 0,
+            label_dir: timeline_label_dir.to_string(),
+            data_dir: timeline_data_dir.to_string(),
+        }
+    }
+
+    pub fn add_snapshot(&mut self, snapshot: Graph) {
+        match &self.curr_snapshot {
+            None => {
+                self.curr_snapshot = Some(snapshot);
+            }
+            Some(_curr_snapshot) => {
+                let is_connected = self.check_connection(&snapshot);
+                self.write_snapshot();
+                self.curr_snapshot = Some(snapshot);
+                if !is_connected {
+                    self.label = 0;
+                    self.write_timeline_label();
+
+                    self.curr_timeline_id += 1;
+                }
+            }
+        }
+        self.curr_snapshot_id += 1;
+    }
+
+    pub fn check_connection(&mut self, other_snapshot: &Graph) -> bool {
+        match &self.curr_snapshot {
+            None => true,
+            Some(curr_snapshot) => {
+                let g1_order = curr_snapshot.get_order();
+                let g2_order = other_snapshot.get_order();
+                let diff = (g1_order as i32 - g2_order as i32).abs();
+                if diff >= 1 {
+                    return false;
+                }
+                true
+            }
+        }
+    }
+
+    pub fn get_curr_snapshot(&self) -> Option<Graph> {
+        match self.curr_snapshot {
+            None => None,
+            Some(ref curr_snapshot) => Some(curr_snapshot.clone()),
+        }
+    }
+
+    pub fn write_snapshot(&mut self) {
+        let curr_snapshot = self.curr_snapshot.as_ref().expect("msg");
+        let mut snapshot_string = take_snapshot_of_graph(curr_snapshot).clone();
+        snapshot_string.push_str("\n");
+
+        let timeline_id = self.curr_timeline_id;
+
+        let filename = format!("{}/{}.txt", self.data_dir, timeline_id);
+        let mut file = OpenOptions::new()
+            .append(true)
+            .create(true)
+            .open(filename)
+            .unwrap();
+
+        if let Err(e) = writeln!(file, "{}", snapshot_string) {
+            eprintln!("Couldn't write to file: {}", e);
```

Comment:
Nitpick: Do we have a defined failure behavior if the user doesn't have these directories (printing a clear error message w instructions, or creating them)? If we don't create them, can we note that we need them in a docstring/readme/centralized place?

Reply:
We can either define that location ourselves or leave it up to the user. If the user enters in a location only root user has access to, the app will probably crash though.

Reply:
Sounds good! Mind adding some detail to the error messages/comments regarding that?

Reply:
Best i can do is:
```rust
if let Err(e) = writeln!(labels_file, "{}", label_text) {
    eprintln!("Couldn't write label to file {}: {}", filename.clone(), e);
}
```
Print which file caused the error followed by the error message. We can't know what caused the error, though the error message should be somewhat straightforward.

Reply:
oh, missed the error in that println. Looks good!